### PR TITLE
WIP of the Kafka consumer wrapper for Subbu

### PR DESCRIPTION
**Pull request by jfim**

### CODE REVIEW

**Comment by jfim** [comment]
This is what the consumer state machine will look when fully implemented.

Kafka consumer state machine

**Comment by mcvsubbu** [comment]
please add this to the design doc in the open source, as well as to the doc
we shared and marked as FINAL Realtime design. thanks

-subbu

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  }
+
+  private abstract class State {
+    private ConsumerState stateValue;
+
+    protected State(ConsumerState stateValue) {
+      this.stateValue = stateValue;
+    }
+
+    abstract void think();
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: what does this do?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+import kafka.message.MessageAndOffset;
+import org.apache.kafka.common.protocol.Errors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * TODO Document me!
+ */
+public class SimpleConsumerWrapper {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: Should we just call it SimpleConsumer? We do have the package name to protect, right?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    }
+  }
+
+  private class ConnectingToBootstrapNode extends State {
+    public ConnectingToBootstrapNode() {
+      super(ConsumerState.CONNECTING_TO_BOOTSTRAP_NODE);
+    }
+
+    @Override
+    public void think() {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: should these be synchronized on SImpleConsumerWrapper? A single wrapper can be only in one state, right?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+  }
+
+  private class ConnectedToBootstrapNode extends State {
+    protected ConnectedToBootstrapNode() {
+      super(ConsumerState.CONNECTED_TO_BOOTSTRAP_NODE);
+    }
+
+    @Override
+    void think() {
+      if (_metadataOnlyConsumer) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: i am tempted to split the metadata only consumer as a separate class with 2 states

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+        // Nothing to do
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: do we want to assert here ?

**Comment by mcvsubbu** [comment]
Overall, I think the approach is workable. Here is what I would prefer:

A state actually does what it is named after (i.e. if it says it is trying to find a bootstrap server, it should be doing that without the need to call 'think'). Alternatively, the state can be named after its pre-condition, and an input can decide what that state does, and which state it goes to next. That would make the state machine more intuitive to understand, IMO. If you can make common methods out of error handlers, that will be great.

Please add unit tests in your final version, mocking kafka as needed (or, overriding it with a fake kafka). Best if unit tests can cover error conditions as well.

In order to support the segment completion protocols, the consumer should have methods like 'start consuming', 'stop consuming and give me the offset', 'catchup to consume upto this offset (exclusive)'. I think you are in the right direction to be able to add these methods, but just wanted to remind you. We can then build segment completion on top of the kafka consumer (and test the segment completion part with a fake kafka consumer wrapper!). The consumer code and the helix state machine code that has been validated is in https://github.com/linkedin/pinot/blob/LLConsumerPrototype/pinot-integration-tests/src/test/java/com/linkedin/pinot/validation/PartitionConsumer.java

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java (hunk 3 above)
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: Yeah I'll synchronize the user-facing methods.

**Comment by jfim** [comment]
I'm not sure how the state machine would work without having a method that's invoked periodically; there's no need to have multiple threads for this and I'd like to avoid synchronization between the Pinot code and some thread that's only there to make sure that on disconnections/connection establishment/leader re-election, we connect to the right node. If you have a good idea of how to make it work without that method and with no external thread, then I'm all ears.

With regards to the naming of the states, they do what the class is named for. Since there's no concurrent or asynchronous execution, the name of the state represents what is going to be done next (eg. if it's in connecting state, the next invocation of `think()` will do the actual connection).

There is no need to have the methods you have mentioned. All that this class does is process requests to consume between offset ranges and ensure that the client is connected to the appropriate Kafka broker for a given partition. The constructs that you have mentioned are something that should be handled in the Pinot realtime consumer, not the Kafka one. This class is only to separate the Kafka-specific details from the Pinot ones.

With regards to unit test, I'll see what I can do without having to create a Kafka consumer factory class and introduce additional complexity.

**Comment by mcvsubbu** [comment]
pls add unit tests at some point.

**Comment by mcvsubbu** [comment]
lgtm otherwise

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java (hunk 3 above)
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: Done

**Pull request merged**